crates/debugger/src/lib.rs

```diff
@@ -0,0 +1,636 @@
+    /// `Response::Complete`.
+    Running,
+    /// Inner body is running in an async task and at a debugger
+    /// callback (or in the initial trampoline waiting for the first
+    /// `Continue`). `Response::Stopped` has been received. Outer
+    /// logic has not sent any commands.
+    Stopped,
+    /// We have sent a command to the inner body and are waiting for a
+    /// response.
+    Queried,
+    /// Inner body is complete (has sent `Response::Finished` and we
+    /// have received it). We may or may not have joined yet; if so,
+    /// the `Option<JoinHandle<...>>` will be `None`.
+    Complete,
+}
+
+/// Message from "outside" to the debug hook.
+///
+/// The `Query` catch-all with a boxed closure is a little janky, but
+/// is the way that we provide access
+/// from outside to the Store (which is owned by `inner` above)
+/// only during pauses. Note that the future cannot take full
+/// ownership or a mutable borrow of the Store, because it cannot
+/// hold this across async yield points.
+///
+/// Instead, the debugger body sends boxed closures which take the
+/// Store as a parameter (lifetime-limited not to escape that
+/// closure) out to this crate's implementation that runs inside of
+/// debugger-instrumentation callbacks (which have access to the
+/// Store during their duration). We send return values
+/// back. Return values are boxed Any values.
+///
+/// If we wanted to make this a little more principled, we could
+/// come up with a Command/Response pair of enums for all possible
+/// closures and make everything more statically typed and less
+/// Box'd, but that would severely restrict the flexibility of the
+/// abstraction here and essentially require writing a full proxy
+/// of the debugger API.
+///
+/// Furthermore, we expect to rip this out eventually when we move
+/// the debugger over to an async implementation based on
+/// `run_concurrent` and `Accessor`s (see #11896). Building things
+/// this way now will actually allow a less painful transition at
+/// that time, because we will have a bunch of closures accessing
+/// the store already and we can run those "with an accessor"
+/// instead.
+enum Command<T: 'static> {
+    Continue,
+    Query(Box<dyn FnOnce(StoreContextMut<'_, T>) -> Box<dyn Any + Send> + Send>),
+}
+
+enum Response {
+    Stopped(DebugRunResult),
+    QueryResponse(Box<dyn Any + Send>),
+    Finished,
+}
+
+struct HandlerInner<T: Send + 'static> {
+    in_rx: Mutex<mpsc::Receiver<Command<T>>>,
+    out_tx: mpsc::Sender<Response>,
+}
+
+struct Handler<T: Send + 'static>(Arc<HandlerInner<T>>);
+
+impl<T: Send + 'static> std::clone::Clone for Handler<T> {
+    fn clone(&self) -> Self {
+        Handler(self.0.clone())
+    }
+}
+
+impl<T: Send + 'static> DebugHandler for Handler<T> {
+    type Data = T;
+    fn handle(
+        &self,
+        mut store: StoreContextMut<'_, T>,
+        event: DebugEvent<'_>,
+    ) -> impl Future<Output = ()> + Send {
+        async move {
```

Comment:
This should be define-able with `async fn handle(...)` and you don't need to worry about `Send`